### ISSUE TITLE
[BUG] Revbayes can't be compiled in c++17 mode

### ISSUE THREAD

**Issue by bredelings** [revbayes/revbayes#238]
GCC and clang on linux only issue warnings, but clang on mac seems to yield hard errors.  This is probably because it is using libc++, whereas clang on linux uses libstdc++ from GCC.

Also, c++17 is the default mode on recent versions of xcode, so this is annoying -- you have to manually select c++14 mode I think.  We'd like to _allow_ c++17 mode now, even if we don't _require_ it until later.

Errors:
* `std::random_shuffle` has been deprecated for a while, and has been removed in c++17 apparently.
* `revlanguage/parser/lex.yy.cpp` is using the deprecated `register` keyword.

**Comment by bredelings** [comment]
The `register` problems can be fixed by (a) regenerating the grammar / lexer and (b) updating our Parser class to match.

See https://github.com/revbayes/revbayes/commit/545c974fef61fc7c4f333a7a45d444dc451ed9ea

**Comment by bredelings** [comment]
The std::random_shuffle problem a bit more complicated.  We need to replace std::random_shuffle with std::shuffle, which takes a random number generator as an additional argument.

See https://clang.llvm.org/extra/clang-tidy/checks/modernize-replace-random-shuffle.html

We could replace `std::random_shuffle(vec.begin(), vec.end())` with `std::shuffle(vec.begin(), vec.end(), std::mt19937(std::random_device()()));`

That is a bit odd because it creates and seeds a new random number generator for each call.

However, it is not used in a large number of places:
```
core/distributions/phylogenetics/tree/birthdeath/SampledSpeciationBirthDeathProcess.cpp:            std::random_shuffle(taxon_idx.begin(), taxon_idx.end());
core/functions/phylogenetics/cladogenetic/DECCladogeneticStateFunction.cpp:            std::random_shuffle( node_child_on.begin(), node_child_on.end() );
core/functions/phylogenetics/cladogenetic/DECCladogeneticStateFunction.cpp:            std::random_shuffle( node_child_on.begin(), node_child_on.end() );
core/functions/phylogenetics/cladogenetic/DECCladogeneticStateFunction.cpp:            std::random_shuffle( node_child_off.begin(), node_child_off.end() );
core/moves/characterhistory/BiogeographyCladogeneticRejectionShiftProposal.h:        std::random_shuffle( contractCharacters.begin(), contractCharacters.end() );
core/moves/characterhistory/BiogeographyCladogeneticRejectionShiftProposal.h://            std::random_shuffle( onCharacters.begin(), onCharacters.end() );
core/moves/characterhistory/BiogeographyCladogeneticRejectionShiftProposal.h:        std::random_shuffle( expandCharacters.begin(), expandCharacters.end() );
core/moves/characterhistory/BiogeographyCladogeneticRejectionShiftProposal.h:        std::random_shuffle( swapCharacters.begin(), swapCharacters.end() );
core/moves/characterhistory/BiogeographyCladogeneticRejectionShiftProposal.h:        std::random_shuffle( leftCharacters.begin(), leftCharacters.end() );
core/moves/characterhistory/BiogeographyCladogeneticRejectionShiftProposal.h:        std::random_shuffle( rightCharacters.begin(), rightCharacters.end() );
core/moves/characterhistory/BiogeographyCladogeneticRejectionShiftProposal.h:    std::random_shuffle( swapCharacters.begin(), swapCharacters.end() );
core/moves/characterhistory/BiogeographyNodeRejectionSampleProposal.h:        std::random_shuffle(idx_on.begin(), idx_on.end());
core/moves/characterhistory/BiogeographyNodeRejectionShiftProposal.h:            std::random_shuffle(idx_on.begin(), idx_on.end());
```

**Comment by bredelings** [comment]
See the `c++17-fixes` branch.  It has the fixes for `register` and the parser, but not for `random_shuffle`.
I've enabled `c++-17` on the branch for the purposes of checking that tests pass when c++17 is enabled...
I'm not proposing that we actually require c++17.  We would need to check that it is supported on most linux clusters first.

**Comment by bredelings** [comment]
This was fixed by #269
